File: iMEGA/Utils/Shared Views/ViewControllers/MEGAQLPreviewController.m
#import "MEGAQLPreviewController.h"
#import "MEGALinkManager.h"
#import "MEGA-Swift.h"

@interface MEGAQLPreviewController () <UIViewControllerTransitioningDelegate, QLPreviewControllerDelegate, QLPreviewControllerDataSource>

@property (nonatomic, strong) NSString *filePath;
@property (nonatomic, strong) NSArray *files;

@end

@implementation MEGAQLPreviewController

- (instancetype)initWithFilePath:(NSString *)filePath {    
    self = [super init];
    
    if (self) {
        _filePath                  = filePath;
        _files                     = nil;
        self.delegate              = self;
        self.dataSource            = self;
        self.title                 = [filePath lastPathComponent];
        self.transitioningDelegate = self;
    }
    
    return self;
}

- (instancetype)initWithArrayOfFiles:(NSArray *)files {
    self = [super init];
    
    if (self) {
        _files                     = files;
        _filePath                  = nil;
        self.delegate              = self;
        self.dataSource            = self;
        self.transitioningDelegate = self;
    }
    
    return self;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.view.backgroundColor = UIColor.pageBackgroundColor;
}

#pragma mark - QLPreviewControllerDataSource

- (NSInteger)numberOfPreviewItemsInPreviewController:(QLPreviewController *)controller {
    if (self.files) {
        return self.files.count;
    } else {
        return 1;
    }
}

- (id <QLPreviewItem>)previewController:(QLPreviewController *)controller previewItemAtIndex:(NSInteger)index {
    if (self.files) {
        self.filePath = [self.files objectAtIndex:index];
    }
    return [NSURL fileURLWithPath:self.filePath];
}

#pragma mark - QLPreviewControllerDelegate

- (BOOL)previewController:(QLPreviewController *)controller shouldOpenURL:(NSURL *)url forPreviewItem:(id <QLPreviewItem>)item {
    dispatch_async(dispatch_get_main_queue(), ^{
        MEGALinkManager.linkURL = url;
        [MEGALinkManager processLinkURL:url];
    });

    return NO;
}


@end

File: iMEGA/Utils/Shared Views/ViewControllers/MEGAQLPreviewController.swift
import MEGADesignToken
import MEGASwift
import QuickLook
import UIKit

final class MEGAQLPreviewController: QLPreviewController, QLPreviewControllerDelegate, QLPreviewControllerDataSource {

    private var files: [String] = []

    @objc init(arrayOfFiles files: [String]) {
        super.init(nibName: nil, bundle: nil)
        self.files = files
        self.delegate = self
        self.dataSource = self
    }
    
    @objc init() {
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        view.backgroundColor = TokenColors.Background.page
    }

    // MARK: - QLPreviewControllerDataSource

    func numberOfPreviewItems(in controller: QLPreviewController) -> Int {
        files.count
    }

    func previewController(_ controller: QLPreviewController, previewItemAt index: Int) -> any QLPreviewItem {
        URL(fileURLWithPath: files[safe: index] ?? "") as (any QLPreviewItem)
    }

    // MARK: - QLPreviewControllerDelegate

    func previewController(_ controller: QLPreviewController, shouldOpen url: URL, for item: any QLPreviewItem) -> Bool {
        DispatchQueue.main.async {
            MEGALinkManager.linkURL = url
            MEGALinkManager.processLinkURL(url)
        }
        return false
    }
}
